go/libkb/merkle_client.go

```diff
@@ -412,21 +694,69 @@ func (mc *MerkleClient) findValidKIDAndSig(root *MerkleRoot) (keybase1.KID, stri
 	return nilKID, "", MerkleClientError{"no known verifying key"}
 }
 
-func (mc *MerkleClient) verifyRoot(ctx context.Context, root *MerkleRoot, seqnoWhenCalled Seqno) error {
+func (mc *MerkleClient) verifySkipSequence(ctx context.Context, ss SkipSequence, thisRoot *MerkleRoot, lastRoot *MerkleRoot) (err error) {
+	defer mc.G().CTrace(ctx, "MerkleClient#verifySkipSequence", func() error { return err })()
+
+	// In this case, the server did not return a skip sequence. It's OK if
+	// the last known root is too old. It's not OK if the last known root is
+	// from after the server starting providing skip pointers.
+	if ss == nil {
+		mc.G().Log.CDebugf(ctx, "| nil SkipSequence")
+		fss := mc.FirstSeqnoWithSkips()
+		if lastRoot == nil {
+			mc.G().Log.CDebugf(ctx, "| lastRoot==nil, so OK")
+			return nil
+		}
+		if fss < 0 {
+			mc.G().Log.CDebugf(ctx, "| no known root with skips, so OK")
+			return nil
+		}
+		if fss > lastRoot.Seqno() {
+			mc.G().Log.CDebugf(ctx, "| lastRoot (%d) is from before first known root with skips (%d), so OK", int(lastRoot.Seqno()), int(fss))
+			return nil
+		}
+		if thisRoot != nil && lastRoot.Seqno() == thisRoot.Seqno() {
+			mc.G().Log.CDebugf(ctx, "| thisRoot is the same as lastRoot (%d), so OK", int(lastRoot.Seqno()))
+			return nil
+		}
+
+		return MerkleClientError{fmt.Sprintf("Expected a skip sequence with last=%d", int(lastRoot.Seqno()))}
+	}
+
+	for i := 0; i < len(ss)-1; i++ {
+		j := i + 1
+		a, b := ss[i].seqno(), ss[j].seqno()
+		mc.G().Log.CDebugf(ctx, "| Checking skip %d->%d", a, b)
+		if a <= b {
+			return MerkleClientError{fmt.Sprintf("Sequence error: %d <= %d", a, b)}
+		}
+		hash := ss[i].skipToSeqno(b)
+		if hash == nil {
+			return MerkleClientError{fmt.Sprintf("Skip missing at %d->%d", a, b)}
+		}
+		if !hashEq(hash, ss[j].shortHash()) {
+			mc.G().Log.CDebugf(ctx, "| Failure in hashes: %s != %s", hash.String(), ss[j].shortHash().String())
+			return MerkleClientError{fmt.Sprintf("Skip pointer mismatch at %d->%d", a, b)}
+		}
+	}
```

Comment:
We know that we put the first and last roots in the list ourselves, so there's no chance that the server's skip list sent us somewhere different from where we _thought_ we were going. But that important security invariant isn't entirely obvious, and if the server's API ever changes somehow to build the full list for us, we'll need to remember to check this property. We could throw some comments in `readSkipSequenceFromAPIRes` to that effect, but I think the safest thing is to just add asserts somewhere in this function that `thisRoot` and `lastRoot` are actually bookending the list. Alternatively, we could move the bookending code into here, and take the responsibility away from `readSkipSequenceFromAPIRes`. I'd be fine with either.